### ISSUE TITLE
New 3.9.0 release 

### ISSUE THREAD

**Issue by navjottomer** [navjottomer/Osclass#83]
New release needs testing, Anything related to 3.9.0 can be discussed here.

**Comment by eurobank** [comment]
Thanks.

Why you suggest to get Bender 3.15 while the latest version is Bender 3.20 ?

**Comment by dev-101** [comment]
US couldn't be installed on localhost during location stage. This is what happened when I tried (out of curiosity) to change drop-down selector during already selected US and installing geo data procedure being initiated.

![2019-10-31_123846](https://user-images.githubusercontent.com/5383059/67944479-d5366700-fbdc-11e9-86d3-543aa292d65c.png)

There was another screen after this one with common flash error message "The selected location could not been installed" (install-functions), but it was brief and I didn't capture it.

Minor detail: we should remove **Readme Feedback Forums** links, also in footer of admin theme which contains them.

**Comment by navjottomer** [comment]
@eurobank  That is a latest release from official bender-repo.
@dev-101 I am on that.

**Comment by dev-101** [comment]
Alternatively, just update osclass link to this repo and forum to webmod's.

**Comment by navjottomer** [comment]
> Alternatively, just update osclass link to this repo and forum to webmod's.

They are already targeting this repo and webmods's forum. Just checked it.

**Comment by eurobank** [comment]
@navjottomer 

The latest Bender is 3.20. I have no idea why 3.15 is on github, still not updated or official.

**Comment by dev-101** [comment]
> > Alternatively, just update osclass link to this repo and forum to webmod's.
> 
> They are already targeting this repo and webmods's forum. Just checked it.

Nope, they are injected here: https://github.com/navjottomer/Osclass/blob/master/oc-admin/themes/modern/functions.php#L60

**Comment by dev-101** [comment]
Yeah, it's a mess with that theme and 3.8.0 release (which was never officially created here).
3.2.0 was released on Market, but not updated on GitHub.
Here's 3.2.0 for comparison.
[bender_3.2.0.zip](https://github.com/navjottomer/Osclass/files/3793711/bender_3.2.0.zip)

**Comment by navjottomer** [comment]
> Nope, they are injected here: https://github.com/navjottomer/Osclass/blob/master/oc-admin/themes/modern/functions.php#L60
> 
> Also, donate form should be removed, too.
my fault only checked installer.php, 
@dev-101  Can you make a pull request with above changes.

**Comment by dev-101** [comment]
Not sure what's the situation here with the branches (develop, hotfix), but Carlos created a bit of a "mess" when stepped-in after Conejo, and pulled all fixed directly into Master branch. He promised in private comm to reset branches and sync them later, but I think that never happened.

In any case, there were fixes (and still some open PRs) in old ("official") Osclass repo there by me and other devs.

Also, once we are done with this release, I think there should be only 1 branch named "prelease" or "develop", and synchronize them with Master.

**Comment by dev-101** [comment]
What I'm trying to say is that current "official" Osclass master branch may contain some fixes which aren't present in here. That should be checked.

**Comment by navjottomer** [comment]
@dev-101 
While creating this fork, I merged a few changes from upstream, we will look into those fixes after this release.
After this release we need to change our branch situation
1. Master branch release will always be production-ready.
2. Develop branch for active development. 
3. prelease branch whenever develop branch is ready for prerelease.
4. Also, other feature specific branches active till the features are not merged in a develop branch.

I hope this is well structured for our need. Currently, our master branch is not clean and that is not good.

**Comment by dev-101** [comment]
Well, ok, as long as you are comfortable maintaining so much branches :)
I am more for simplicity, though, but ok.

**Comment by eurobank** [comment]
Keeping the 3.8 db, i uploade 3.9rc and i got a message to upgrade. 
I did, and then i got an OK message BUT with an old 3.8 message.

![2019-10-31 15_02_51-Osclass 3 9 0 » Test Osclass Site - Osclass](https://user-images.githubusercontent.com/4467158/67949109-c5744e00-fbef-11e9-90b6-070227e141d7.png)

**Comment by eurobank** [comment]
Adding a location also doesn't work. Should it ?

Some notices ...

>  
> [31-Oct-2019 14:07:10 Europe/Madrid] PHP Notice:  Undefined index: name in /home/oc-admin/controller/settings/locations.php on line 463
> [31-Oct-2019 14:07:10 Europe/Madrid] PHP Notice:  Undefined index: name in /home/oc-admin/controller/settings/locations.php on line 463
> [31-Oct-2019 14:07:10 Europe/Madrid] PHP Notice:  Undefined index: name in /home/oc-admin/themes/modern/settings/locations.php on line 435
> [31-Oct-2019 14:07:10 Europe/Madrid] PHP Notice:  Undefined index: name in /home/oc-admin/themes/modern/settings/locations.php on line 435

**Comment by dev-101** [comment]
oc-includes\osclass\installer\readme.php also contains a lot of old osclass.org domain references.

**Comment by dev-101** [comment]
> Keeping the 3.8 db, i uploaded 3.9rc and i got a message to upgrade.
> I did, and then i got an OK message BUT with an old 3.8 message. With an Osclass forums link for more help.

This is because of old file oc-includes\osclass\assets\release.notes.php

**Comment by eurobank** [comment]
I still see a "freeze" when i login to admin OR to the home page for the 1st time.

**Comment by eurobank** [comment]
Those must be removed from PHP files:

oc-load.php

1.331	$message = 'There doesn\'t seem to be a <code>config.php</code> file. Osclass isn\'t installed. <a href="http://forums.osclass.org/">Need more help?</a></p>';

3.267	$message = 'Osclass isn\'t installed. <a href="http://forums.osclass.org/">Need more help?</a></p>';

--

functions.php

2.501	<a title="<?php _e('Forums'); ?>" href="http://forums.osclass.org/" target="_blank"><?php _e('Forums'); ?></a> &middot;

--

emails.php

58.462	$body .= '<p><a href="http://forums.osclass.org/">http://forums.osclass.org/</a></p>';
58.490	$body .= '<p><a href="http://forums.osclass.org/">http://forums.osclass.org/</a></p>';

--

install.php

12.197	<a href="http://forums.osclasscommunity.com/" target="_blank"

--

install-locations.php

3.960	__( 'Remember that for any doubts you might have you can consult our <a href="%1$s">documentation</a>, <a href="%2$s">forum</a> or <a href="%3$s">blog</a>.' ) , 'http://doc.osclass.org/' , 'http://forums.osclass.org/' , 'http://blog.osclass.org/' );

--

upgrade-funcs.php

2.136	.= '<br/><br/>' . sprintf( __( "These errors could be false-positive errors. If you're sure that is the case, you can <a href=\"%s\">continue with the upgrade</a>, or <a href=\"http://forums.osclass.org/\">ask in our forums</a>."), $skip_db_link);

**Comment by eurobank** [comment]
Also there are tens of links like the above in PO and MO files.

**Comment by navjottomer** [comment]
@eurobank Please keep this to discussion only, report your finding at #84 .

**Comment by webmods-croatia** [comment]
@dev101 @navjottomer https://github.com/navjottomer/Osclass/pull/87

**Comment by eurobank** [comment]
Please fix the freeze/delay still available in the rc. 

Due to the "ajax.php" file for some reason, i'm not a programmer and i can't tell why exactly. Commenting out all the calls in that file to "osclass.org" solves the delay.

**Comment by webmods-croatia** [comment]
@eurobank Do you maybe have Google maps or Google analytics plugins installed? I think they caused some problems because they use different update URLs.

**Comment by eurobank** [comment]
@webmods-croatia Nope, no plugins.

It is AJAX.PHP doing that. I already wrote that commenting out the calls in update code, solves it.

**Comment by eurobank** [comment]
And applying this

https://docs.osclasscommunity.com/removing-market/remove-calls-to-osclass/stop-checking-for-updates-from-osclass  (the ajax.part)

to ajax.php solves it.

**Comment by webmods-croatia** [comment]
@eurobank Wow, my bad. I made the guide and I did the changes. Must have missed it. I will make a PR ASAP.

**Comment by webmods-croatia** [comment]
@eurobank https://github.com/navjottomer/Osclass/pull/88

**Comment by eurobank** [comment]
Maybe a silly question but what happens with OLD installations and this:
`
 ,('osclass', 'marketURL', 'https://market.osclass.org/api/v3/', 'STRING')`

that is removed/replaced now in 3.9 rc ?

**Comment by webmods-croatia** [comment]
@eurobank All calls to Osclass domains are removed but code is kept for upgrade system in future.

**Comment by eurobank** [comment]
@navjottomer 

What i mean is that right now i have the above in my Database (if i understand correctly). You changed that for NEW installations. OLD installations that will update to 3.9 will still have it unchanged.

Does it matter for old ones ?

**Comment by webmods-croatia** [comment]
@eurobank That preference in database does nothing. It's just Market URL that's supposed to be used for updating but we removed updating in 3.9. No need to worry.

**Comment by eurobank** [comment]
@web-media OK, i guess at some point it should be changed somehow, in order to support future updates ... when that is available .....

**Comment by navjottomer** [comment]
We are in the transition phase right now, Next goal after this build is to enable auto-update in osclass. Migrating old library to a modern one and remove some of them if they are not necessary.

**Comment by eurobank** [comment]
Maybe this could be fixed/changed. It will confuse people.

![2019-11-04 15_34_07-Window](https://user-images.githubusercontent.com/4467158/68124743-b60b4280-ff18-11e9-89da-ce1c623d96bb.png)

**Comment by navjottomer** [comment]
> Maybe this could be fixed/changed. It will confuse people.
> 
> 
Waiting for a link-replace branch merge to avoid conflict, will be fixed.

**Comment by web-media** [comment]
Hy all .. glad  to be here ... to be involved .

**Comment by webmods-croatia** [comment]
@web-media Oh, hello there! Not sure if above is sarcasm or not but welcome anyways. ;)

**Comment by webmods-croatia** [comment]
@navjottomer 

> Waiting for a link-replace branch merge to avoid conflict, will be fixed.

Will you fix it yourself or you want me to do so?

**Comment by navjottomer** [comment]
> @navjottomer
> 
> > Waiting for a link-replace branch merge to avoid conflict, will be fixed.
> 
> Will you fix it yourself or you want me to do make a PR?

I have it on my list. Thanks!

**Comment by eurobank** [comment]
I asked before and got no answer. 

Does the installation or addition of a Location (from your repo here) work in the admin dashboard? 

Should it work ?

**Comment by navjottomer** [comment]
> I asked before and got no answer.
> 
> Does the installation or addition of a Location (from your repo here) work in the admin dashboard?
> 
> Should it work?

Yes, they work now in the release branch.

**Comment by navjottomer** [comment]
@eurobank @webmods-croatia @dev-101 
I think the pre-release branch is stable enough for RC2. What's your opinion?

**Comment by webmods-croatia** [comment]
@navjottomer I didn't have time to do proper testing myself but all looks fine.

**Comment by eurobank** [comment]
@navjottomer 

Yes i haven't found anything wrong yet.

**Comment by eurobank** [comment]
But there is not enough people testing i'm afraid.

**Comment by webmods-croatia** [comment]
I will announce new version in hope to get some testers of forum and Market.

**Comment by navjottomer** [comment]
I haven't tagged a new release, wait for it.

**Comment by eurobank** [comment]
What that we all havent done is some test WITH the official plugins of Osclass. Not the 3rd party, the ones made by Osclass.

**Comment by webmods-croatia** [comment]
No hooks or filters are touched so all plugins should work.

**Comment by eurobank** [comment]
@webmods-croatia Still i will add a few of the most popular plugins and check. At least catch any big problem, since you can't really fully test without users poking around ...

**Comment by navjottomer** [comment]
Although I tried to avoid any incompatibility to minimum v3.7.0 version of osclass this can't be guaranteed. I tried to test plugins locally ( Many free and 3rd party plugins ) and didn't found any issue.
But need more testing for the full stable release.
I merged everything to master because it wasn't clean before, even develop branch was more stable than the master branch.

**Comment by eurobank** [comment]
So everything is on master now ? Is that the RC2 ?

**Comment by eurobank** [comment]
There is still a link in "locations.php" (oc-admin/controller/settings). 

Maybe it is nothing. Can somebody check it ?

**Comment by eurobank** [comment]
Spend a few hours with the RC2 (from master). Also some trusted users helped. 

I installed several Osclass plugins, most of them official from Osclass team. Also i installed the most popular theme from Osclass Wizards.

I found no errors with the Osclass team made plugins, no errors also with Osclass Wizards theme.

Also no errors with 4 popular plugins from 3rd party, that i can't live without. 

The only problem i had was from Spam protection plugin (Liath's) that crashed hard in my debug.log, something about phpmailer. This is a stopper for me but not for the release of cource.

**Comment by webmods-croatia** [comment]
> There is still a link in "locations.php" (oc-admin/controller/settings).
> 
> Maybe it is nothing. Can somebody check it ?

@navjottomer There is old geo code in that file (_$data_sql = osc_file_get_contents('https://geo.osclass.org/newgeo.download.php?action=country&term=' . urlencode($countryCode) );_) that is not used. Should we keep it or not?

**Comment by webmods-croatia** [comment]
@eurobank I installed spam protection all all seems fine. May you share debug.log? What should I do to reproduce the problem?

**Comment by webmods-croatia** [comment]
@eurobank I installed spam protection and all seems fine. May you share debug.log? What should I do to reproduce the problem?

**Comment by navjottomer** [comment]
@webmods-croatia 
> > There is still a link in "locations.php" (oc-admin/controller/settings).
> @navjottomer There is old geo code in that file (_$data_sql = osc_file_get_contents('https://geo.osclass.org/newgeo.download.php?action=country&term=' . urlencode($countryCode) );_) that is not used. Should we keep it or not?

Remove it not needed.

**Comment by eurobank** [comment]
@webmods-croatia I will have someone email you directly the problem. I don't want to polute here.

**Comment by webmods-croatia** [comment]
> @webmods-croatia I will have someone email you directly the problem. I don't want to polute here.

Or go to http://osclass.discourse.group/

**Comment by webmods-croatia** [comment]
@navjottomer https://github.com/navjottomer/Osclass/pull/96

**Comment by eurobank** [comment]
For me adding a new Location via admin dashboard doesn't work at all.

Any country i try, the list is NOT populated.

![2019-11-06 13_32_13-Window](https://user-images.githubusercontent.com/4467158/68294993-0a89fb80-009a-11ea-96fe-69f798c0c399.png)

**Comment by webmods-croatia** [comment]
@eurobank Add country is for manual adding oc countries. Click "Import new".

**Comment by eurobank** [comment]
@webmods-croatia 

Add country is NOT for manually, it was populating from Geo db.

**Comment by webmods-croatia** [comment]
@eurobank 

You could add a country manually if you wished so with the same form...

Import location is now for importing.

**Comment by eurobank** [comment]
@webmods-croatia I don't follow.

So it is something that has changed permantently  or doesn't work for now ?

Is that working when a NEW installation is done ? I havem't checked that yet, but i'm going to.

**Comment by webmods-croatia** [comment]
@eurobank To import a location from our database click "Import new" and pick from list.

You can pick a location on install.

**Comment by eurobank** [comment]
After adding a location, i try to add a second one. And i fail.

After deleting all locations, i was able to do that and add multiple countries.

Also deleting a location with a lot of data (USA for example or France) it takes 2-3 minutes. I don't recall that to happen previously.

**Comment by eurobank** [comment]
I was able to replicate the issue.

1. Delete all locations.

2. Add Andorra

3. Try to add an second.

**Comment by web-media** [comment]
![Capture](https://user-images.githubusercontent.com/5124319/68346389-494c9f80-00f4-11ea-9d3f-70b309383cfd.PNG)
 On install Countries (locationSql )  should be sorted by Country name

**Comment by eurobank** [comment]
I don't understand fully but they are not sorted. Angolla is 1st in the selectio box:

![2019-11-07 01_54_10-Window](https://user-images.githubusercontent.com/4467158/68348036-88352200-0101-11ea-8a05-912c57f5ec70.png)

but see the html also:

![2019-11-07 01_55_12-Window](https://user-images.githubusercontent.com/4467158/68348099-b3b80c80-0101-11ea-9a6f-4b7776008abc.png)

**Comment by navjottomer** [comment]
> I was able to replicate the issue.
> 
> 1. Delete all locations.
> 2. Add Andorra
> 3. Try to add an second.

I couldn't replicate it, location importing working fine for me.

**Comment by eurobank** [comment]
@navjottomer Let me know if there is something i can do more to desolve this. Debug log is empty.

**Comment by eurobank** [comment]
Here is what exactly happens.

https://drive.google.com/open?id=1wrhmY79-25EvfZMiXASwPqE_sHxWY4Dg

**Comment by navjottomer** [comment]
BTW, I got it now.
@eurobank Discuss it here for future discussion because this thread is getting long.
https://osclass.discourse.group/t/osclass-3-9-0-release-related-discussion/36

**Comment by eurobank** [comment]
@navjottomer Well spliting things here and there and changing all the time, isn't gonna help.

Please do not get me wrong, but this is how it is.

**Comment by navjottomer** [comment]
Sorry, my fault! 
I don't want to indulge in any debate, I should focus on my job.

**Comment by eurobank** [comment]
@navjottomer No debate at all. Not from me. But if this become complicated and hard to follow, it will be left only for developers. 

And it is you that posted "New release needs testing, Anything related to 3.9.0 can be discussed here.". 

But now you post differently.

Anyways, i will open new issues next time when i find something. Better that way.

**Comment by webmods-croatia** [comment]
Sorry for all the problems. I will hopefully be able to check this today.

**Comment by navjottomer** [comment]
@webmods-croatia It is fixed now.

**Comment by webmods-croatia** [comment]
@navjottomer What was the issue?
Did you fix ordering by name as well?

**Comment by navjottomer** [comment]
### Ordering
> Ordering by name is fixed in dir2json list. https://github.com/navjottomer/Osclass-Extras/commit/b6505e588c3871950ffa094cfba0eeda49b362b2
### Location bug:
https://github.com/navjottomer/Osclass/commit/d64914dc8b599f80502dbddb8605eafb82eb7ee0 
The issue was in the condition.

**Comment by eurobank** [comment]
The import locations problem seems fixed now, with the master copy i installed.

**Comment by webmods-croatia** [comment]
@navjottomer What's remaining to be done for release?

**Comment by navjottomer** [comment]
@webmods-croatia I think just need to update changelog and release it.

**Comment by webmods-croatia** [comment]
@navjottomer Then let's do it! I'll update my site to 3.9.0.

**Comment by eurobank** [comment]
No do not release anything. Make please a final RC, FREEZE IT and wait at least a week or so.

Do not rush.

**Comment by Tangol** [comment]
> The only problem i had was from Spam Protection plugin (Liath's) that crashed hard in my debug.log, something about phpmailer. This is a stopper for me but not for the release of cource.

I also got a warning with this plugin:
`PHP Warning:  in_array() expects parameter 2 to be array, boolean given in C:\Bitnami\wampstack-7.0.33-0\apache2\htdocs\stock\oc-content\plugins\spamprotection\classes\class.spamprotection.php on line 1349`

And here is the code:
```
$table = unserialize($this->_get('sp_ipban_table'));
        if (in_array($ip, $table)) {
            return true;
        }
```
What could be the issue?

**Comment by navjottomer** [comment]
@Tangol this is just a warning, It shouldn't be a reason for osclass crash.

**Comment by navjottomer** [comment]
BTW I think prelease branch is stable and can be merged and we can release 3.9.0

**Comment by webmods-croatia** [comment]
@navjottomer So are we doing it. :)

**Comment by navjottomer** [comment]
@webmods-croatia It is released now. 
closing this issue too.

**Comment by buninsan** [comment]
when installing any plugin I get an error
The plugin generated 1 characters of unexpected output during the installation. Output: "
"
I also see an error in the template
/oc-includes/osclass/helpers/hSearch.php
function osc_search_url($params = null) {
function osc_search_url($params = array()) {

**Comment by dev-101** [comment]
I also noticed that since moving dependencies to /vendor/ dir, some plugins of mine are effectively broken. Compatibility solution for this is a bit ugly, but necessary if we wish to keep everything in the works for those who wish to transition from 3.8.0.

**Comment by navjottomer** [comment]
@buninsan 
> when installing any plugin I get an error
> The plugin generated 1 characters of unexpected output during the installation. Output: "
> "
> I also see an error in the template
> /oc-includes/osclass/helpers/hSearch.php
> function osc_search_url($params = null) {
> function osc_search_url($params = array()) {
Not helpful! Plugin name? Did you checked permissions?

@dev-101 Can you share more information.

BTW, we shouldn't discuss this on a closed issue thread. Better if we move this discussion to https://osclass.discourse.group

**Comment by dev-101** [comment]
Sorry, haven't got time to register there yet. For example, one is using HTML Purifier custom instance, so the loading will fail and there was no safeguard against such event (simply, I couldn't predict it, and since old Market is down, releasing new version is impossible).

```require_once LIB_PATH . 'htmlpurifier/HTMLPurifier.auto.php';```

The ugly solution is to place loader file that will point to the new location.

**Comment by buninsan** [comment]
> Sorry, haven't got time to register there yet. For example, one is using HTML Purifier custom instance, so the loading will fail and there was no safeguard against such event (simply, I couldn't predict it, and since old Market is down, releasing new version is impossible).
> 
> `require_once LIB_PATH . 'htmlpurifier/HTMLPurifier.auto.php';`
> 
> The ugly solution is to place loader file that will point to the new location.

require_once LIB_PATH . 'vendor/ezyang/htmlpurifier/library/HTMLPurifier.auto.php';

**Comment by navjottomer** [comment]
@buninsan in new version require_once is not needed, external libs are autoloaded.

**Comment by buninsan** [comment]
Thank you, your advice helped me, this solution works:
//	require_once LIB_PATH . 'htmlpurifier/HTMLPurifier.auto.php';
